flang/lib/Lower/OpenMP.cpp

```diff
@@ -75,6 +76,43 @@ genOMP(Fortran::lower::AbstractConverter &absConv,
       standaloneConstruct.u);
 }
 
+static void
+genOMP(Fortran::lower::AbstractConverter &absConv,
+       Fortran::lower::pft::Evaluation &eval,
+       const Fortran::parser::OpenMPBlockConstruct &blockConstruct) {
+  const auto &blockDirective =
+      std::get<Fortran::parser::OmpBeginBlockDirective>(blockConstruct.t);
+  const auto &parallelDirective =
+      std::get<Fortran::parser::OmpBlockDirective>(blockDirective.t);
+  if (parallelDirective.v == llvm::omp::OMPD_parallel) {
+    auto &firOpBuilder = absConv.getFirOpBuilder();
+    auto currentLocation = absConv.getCurrentLocation();
+    auto insertPt = firOpBuilder.saveInsertionPoint();
+    llvm::ArrayRef<mlir::Type> argTy;
+    mlir::ValueRange range;
+    mlir::OperationState opState(currentLocation,
+                                 mlir::omp::ParallelOp::getOperationName());
+    // FIXME: This can be done in a better way ?
+    std::array<int32_t, 6> operandSegmentSizes = {0, 0, 0, 0, 0, 0};
+    opState.addAttribute(mlir::omp::ParallelOp::getOperandSegmentSizeAttr(),
+                         firOpBuilder.getI32VectorAttr(operandSegmentSizes));
+    // build the operation with the specified state.
+    mlir::omp::ParallelOp::build(firOpBuilder, opState, argTy, range);
+    // create and insert the operation.
+    mlir::Operation *op = firOpBuilder.createOperation(opState);
+    auto parallelOp = dyn_cast<mlir::omp::ParallelOp>(op);
+    parallelOp.getRegion().push_back(new Block{});
+    auto &block = parallelOp.getRegion().back();
+    firOpBuilder.setInsertionPointToStart(&block);
+    // `parallel` region has an implicit barrier at the very end.
+    firOpBuilder.create<mlir::omp::BarrierOp>(currentLocation);
```

Comment:
Should this be added here or when we translate to LLVM IR?

Reply:
I'm not sure but, could you please explain why this fits better there in translation phase ?
IMHO we should create/insert barrier here it self when we complete the operation creation. Instead of doing it very late.
If we follow the approach of late translation then. In the situation of `NoWait` clause, then this decision of putting barrier is also getting delayed till we actually perform the translation.

Reply:
I think it's possibly better to be here as when writing optimisation passes you'll have the information that there is a barrier here. That might possibly be useful knowledge for some optimisations. Especially since the barrier won't be there if you have written nowait.

Reply:
It seems a bit odd here.
Is the implicit barrier in the kmpc_fork_call?

Reply:
Yeah sort of confusion here, perhaps you can help clearing this.
For an empty parallel pragma:
`
#pragma omp parallel
{}
`
I didn't notice "__kmpc_barrier" in IR generated from clang.
I concur @kiranchandramohan is correct, perphaps implicit barrier is correct.  @kiranchandramohan can you confirm this, so that I can correct this.
Thanks!

Reply:
I can confirm the presence of implicit barrier calls (after inspecting omp runtime code and in a debug session). So you can remove this explicit barrier insertion.

Note: Calls are to __kmp_join_barrier via
1) __kmpc_fork_call->__kmp_join_call->__kmp_internal_join->__kmp_join_barrier
2) (indirectly from kmpc_fork_call) thread_start->start_thread->__kmp_launch_worker->__kmp_launch_thread->__kmp_join_barrier.

Reply:
Thanks! @kiranchandramohan for confirming our understanding and sharing this info.
I've removed these explicit calls.

Reply:
I wonder if this might be something we want to re-examine later, for the reason I posted above. Knowing when there will be a barrier and when there will not in an optimisation pass seems like very useful information to me. Then again, it may not ever actually turn out to be useful and I may be wrong about that!